clients/drcachesim/scheduler/scheduler_impl.cpp

```diff
@@ -2415,27 +2380,79 @@ scheduler_impl_tmpl_t<RecordType, ReaderType>::pick_next_input(output_ordinal_t
     }
     // We can't easily place these stats inside set_cur_input() as we call that to
     // temporarily give up our input.
-    update_switch_stats(output, prev_index, index);
+    on_context_switch(output, prev_index, index);
     set_cur_input(output, index);
     return res;
 }
 
 template <typename RecordType, typename ReaderType>
 void
-scheduler_impl_tmpl_t<RecordType, ReaderType>::update_switch_stats(
+scheduler_impl_tmpl_t<RecordType, ReaderType>::on_context_switch(
     output_ordinal_t output, input_ordinal_t prev_input, input_ordinal_t new_input)
 {
+    bool do_inject_switch_seq = false;
     if (prev_input == new_input)
         ++outputs_[output].stats[memtrace_stream_t::SCHED_STAT_SWITCH_NOP];
     else if (prev_input != sched_type_t::INVALID_INPUT_ORDINAL &&
-             new_input != sched_type_t::INVALID_INPUT_ORDINAL)
+             new_input != sched_type_t::INVALID_INPUT_ORDINAL) {
         ++outputs_[output].stats[memtrace_stream_t::SCHED_STAT_SWITCH_INPUT_TO_INPUT];
-    else if (new_input == sched_type_t::INVALID_INPUT_ORDINAL)
+        do_inject_switch_seq = true;
+    } else if (new_input == sched_type_t::INVALID_INPUT_ORDINAL)
         ++outputs_[output].stats[memtrace_stream_t::SCHED_STAT_SWITCH_INPUT_TO_IDLE];
     else {
         ++outputs_[output].stats[memtrace_stream_t::SCHED_STAT_SWITCH_IDLE_TO_INPUT];
         // Reset the flag so we'll try to steal if we go idle again.
         outputs_[output].tried_to_steal_on_idle = false;
+        do_inject_switch_seq = true;
+    }
+
+    // We want to insert the context switch sequence on input-to-input and
+    // idle-to-input cases. This is a better control point to do that than
+    // set_cur_input. Here we get the stolen input events too, and we don't have
+    // to filter out the init-time set_cur_input cases.
+    // RFC: Was there any other benefit to doing this in set_cur_input before?
+    // If this logic is kept a part of set_cur_input, when USE_INPUT_ORDINALS is
+    // set, it is harder to differentiate between idle-to-input events and start
+    // of output; this is because get_instruction_ordinal() does not return the
+    // intended global value in that case (it returns the input-local value
+    // which is also not adjusted for the scheduler read-ahead, so is non-zero).
+    if (do_inject_switch_seq) {
+        if (inputs_[new_input].pid != INVALID_PID) {
+            insert_switch_tid_pid(inputs_[new_input]);
+        }
+        if (!switch_sequence_.empty()) {
+            switch_type_t switch_type = sched_type_t::SWITCH_INVALID;
+            if (prev_input == sched_type_t::INVALID_INPUT_ORDINAL ||
```

Comment:
This check wasn't there before: is this a behavior change?  If the start of a thread is already not reaching here: when is this set -- on idle to input transitions?  Add a comment?

Reply:
After offline discussion: changed to XXX comment noting that we may want to improve this heuristic.